pkg/uploader/provider/kopia.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright The Velero Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kopia/kopia/snapshot/snapshotfs"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vmware-tanzu/velero/pkg/uploader"
+	"github.com/vmware-tanzu/velero/pkg/uploader/kopia"
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	repokeys "github.com/vmware-tanzu/velero/pkg/repository/keys"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo/service"
+)
+
+//BackupFunc mainly used to make testing more convenient
+var BackupFunc = kopia.Backup
+var RestoreFunc = kopia.Restore
+
+//kopiaProvider recorded info related with kopiaProvider
+type kopiaProvider struct {
+	bkRepo        udmrepo.BackupRepo
+	credGetter    *credentials.CredentialGetter
+	uploader      *snapshotfs.Uploader
+	restoreCancel chan struct{}
+	log           logrus.FieldLogger
+}
+
+//NewKopiaUploaderProvider initialized with open or create a repository
+func NewKopiaUploaderProvider(
+	ctx context.Context,
+	credGetter *credentials.CredentialGetter,
+	bsl *velerov1api.BackupStorageLocation,
+	log logrus.FieldLogger,
+) (Provider, error) {
+	kp := &kopiaProvider{
+		log:        log,
+		credGetter: credGetter,
+	}
+	//repoUID which is used to generate kopia repository config with unique directory path
+	repoUID := string(bsl.GetUID())
+	repoOpt, err := udmrepo.NewRepoOptions(
+		udmrepo.WithPassword(kp, ""),
+		udmrepo.WithConfigFile("", repoUID),
+		udmrepo.WithDescription("Initial kopia uploader provider"),
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error to get repo options")
+	}
+
+	repoSvc := service.Create(log)
+	log.WithField("repoUID", repoUID).Info("Opening backup repo")
+
+	kp.bkRepo, err = repoSvc.Open(ctx, *repoOpt)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to find kopia repository")
+	}
+	return kp, nil
+}
+
+//CheckContext check context status periodically
+//check if context is timeout or cancel
+func (kp *kopiaProvider) CheckContext(ctx context.Context) {
+	for {
```

Comment:
The for loop should be stopped once the backup/restore is finished

Reply:
yes, because it is used in RunBackup or RunRestore as one goroutine, once RunBackup or RunRestore is finished. the goroutine will exit

Reply:
Seems the for loop isn't needed. By removing the `default` section, we can use select without the for loop